File: src/test/java/com/example/scaffold/IntegrationTest.java
package com.example.scaffold;

import org.junit.jupiter.api.*;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.*;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.web.util.UriComponentsBuilder;
import org.testcontainers.shaded.com.google.common.collect.ImmutableMap;

import javax.annotation.Resource;
import java.net.URI;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

@Tag("slow")
@ExtendWith({TestContainersInitializer.class, SpringExtension.class, MocoInitializer.class})
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ActiveProfiles("unittest")
public abstract class IntegrationTest {

    protected @Resource TestRestTemplate testRestTemplate;
    private @Resource TestContainersInitializer testContainersInitializer;
    protected String authorization;

    @BeforeEach
    void testIsContainerRunning() {
        assertThat(testContainersInitializer.isRunning()).isTrue();
    }

    protected TestResponse get(String urlTemplate, Object... vars) {
        return get(urlTemplate, ImmutableMap.of(), vars);
    }

    protected TestResponse put(String urlTemplate, Object body, Object... vars) {
        return exchange(getRequest(urlTemplate, body, vars, HttpMethod.PUT));
    }

    protected TestResponse post(String urlTemplate, Object body, Object... vars) {
        return exchange(getRequest(urlTemplate, body, vars, HttpMethod.POST));
    }

    protected TestResponse get(String urlTemplate, Map<String, Object> params, Object... vars) {
        UriComponentsBuilder builder = UriComponentsBuilder.fromUriString(urlTemplate);
        params.forEach(builder::queryParam);
        URI uri = builder.buildAndExpand(vars).encode().toUri();
        RequestEntity<?> request = new RequestEntity<>(requestHeader(), HttpMethod.GET, uri);
        return exchange(request);
    }

    protected TestResponse delete(String url, Map<String, Object> params) {
        RequestEntity<Map<String, Object>> httpEntity = new RequestEntity<>(params, requestHeader(), HttpMethod.DELETE, URI.create(url));
        return exchange(httpEntity);
    }

    private TestResponse exchange(RequestEntity<?> request) {
        ResponseEntity<String> exchange = testRestTemplate.exchange(request, String.class);
        return new TestResponse(exchange);
    }

    private RequestEntity<?> getRequest(String urlTemplate, Object body, Object[] vars, HttpMethod put) {
        URI uri = UriComponentsBuilder.fromUriString(urlTemplate).buildAndExpand(vars).encode().toUri();
        return new RequestEntity<>(body, requestHeader(), put, uri);
    }

    private HttpHeaders requestHeader() {
        HttpHeaders headers = new HttpHeaders();
        addAuthorization(headers);
        return headers;
    }

    private void addAuthorization(HttpHeaders headers) {
        headers.add("Authorization", authorization);
    }

}

File: src/test/kotlin/com/qomolangma/IntegrationTest.kt
package com.qomolangma

import com.qomolangma.frameworks.domain.core.GlobalClock
import com.qomolangma.frameworks.test.container.WithTestContainers
import com.qomolangma.frameworks.test.moco.MocoContainer
import com.qomolangma.frameworks.test.web.GlobalTestContext
import com.qomolangma.frameworks.test.web.Rest
import com.qomolangma.frameworks.security.core.AccessToken
import com.qomolangma.frameworks.security.core.User
import com.qomolangma.frameworks.security.core.UserIdentityOwner
import com.qomolangma.frameworks.security.core.Users
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import javax.annotation.Resource

@ActiveProfiles("unittest")
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@WithTestContainers(containers = [MocoContainer::class])
abstract class IntegrationTest : Rest() {
    @Resource
    private val users: Users? = null

    @Resource
    private val accessTokens: User.AccessTokens? = null
    private var user: User? = null
    private var accessToken: AccessToken? = null

    protected fun removeAccessToken() {
        users!!.clear()
        accessTokens!!.clear()
    }

    @BeforeEach
    fun setupAccessToken() {
        GlobalTestContext.context().ifPresent {
            user = User.create(
                UserIdentityOwner(it.userContext()),
                it.securityContext().grantedAuthorities()
            )
            accessToken = AccessToken(
                user!!.id(),
                it.securityContext().id(),
                GlobalClock.now()
            )
            accessTokens!!.add(accessToken!!)
            users!!.add(user!!)
        }
    }

    @AfterEach
    fun teardownAccessToken() {
        removeAccessToken()
    }
}
